src/serde/ser.rs

```diff
@@ -1,12 +1,75 @@
+use std::fmt::Arguments;
 use std::io;
 use std::io::Cursor;
+use std::io::ErrorKind;
+use std::io::IoSlice;
+use std::io::Write;
 
 use super::write_atom::write_atom;
 use crate::allocator::{NodePtr, SExp};
 use crate::node::Node;
 
 const CONS_BOX_MARKER: u8 = 0xff;
 
+pub struct LimitedWriter<W: io::Write> {
+    inner: W,
+    limit: i64,
+}
+
+impl<W: io::Write> LimitedWriter<W> {
+    pub fn new(w: W, limit: i64) -> LimitedWriter<W> {
+        LimitedWriter { inner: w, limit }
+    }
+
+    pub fn into_inner(self) -> W {
+        self.inner
+    }
+}
+
+impl<W: io::Write> Write for LimitedWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if (self.limit - buf.len() as i64) < 0 {
+            return Err(ErrorKind::OutOfMemory.into());
+        }
+        let written = self.inner.write(buf)?;
+        self.limit -= written as i64;
+        Ok(written)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
+        let mut ret: usize = 0;
+        for b in bufs {
+            let written = self.write(b)?;
+            ret += written;
+            if written < b.len() {
+                break;
+            }
+        }
+        Ok(ret)
+    }
+    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+        self.limit -= buf.len() as i64;
```

Comment:
I may be wrong, but I don't think you have to implement anything besides `write` & `flush`. The defaults should probably be just fine.

Reply:
oh. I didn't realize that. I'll take a look!